src/llm/io_processing/phi4/tool_parser.cpp

```diff
@@ -86,8 +96,223 @@ void Phi4ToolParser::parse(ParsedOutput& parsedOutput, const std::vector<int64_t
 }
 
 std::optional<rapidjson::Document> Phi4ToolParser::parseChunk(const std::string& chunk, ov::genai::GenerationFinishReason finishReason) {
-    // Not implemented
-    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Phi4OutputParser::parseChunk is not implemented");
+    /* 
+    Phi4 with vLLM template produces tool calls in the format:
+    functools[{"name": [function name], "arguments": [function arguments as JSON]}, ...]
+
+    Due to the tool call format used by Phi4, we need to track the state of parsing more closely.
+    We have four states:
+    1) AWAITING_START_TAG - we are waiting for the "functools" tag to appear in the chunk
+    2) AWAITING_TOOL_CALLS_OPENING_BRACKET - we have seen "functools" but are waiting for the opening bracket of the array
+    3) AWAITING_TOOL_CALL_OPENING_BRACE - we have seen the opening bracket of the array but are waiting for the opening brace of the next tool call object
+    4) PROCESSING_TOOL_CALL - we are processing the tool call object
+
+    To avoid missing any generated content, we use unprocessedBuffer to store any output that is not used in the current state, but might be relevant in the next state.
+    Since tools calls in the array are separated by commas we also need to track when the tool call object ends (no special tag for that).
+    Next challenge, common for all parsers, is to return arguments as string even though model generates them as JSON.
+    We address this by escaping double quotes and adding opening quote at the beginning of arguments and closing quote at the end of arguments.
+    */
+    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Phi4ToolParser::parseChunk called with chunk: '{}', finishReason: {}", chunk, static_cast<int>(finishReason));
+    if (chunk.empty()) {
+        SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Received empty chunk for Phi4ToolParser");
+        return std::nullopt;
+    }
+
+    // We merge unprocessedBuffer from previous calls with the current chunk to avoid losing any content
+    std::string modifiedChunk = unprocessedBuffer + chunk;
+    unprocessedBuffer.clear();
+
+    // Before we have 'arguments' in the JSON, we do not want to process both key and value in the same call due to special handling of arguments value.
+    // We look for colon after 'arguments' key and move everything after it to unprocessedBuffer to be processed in the next call.
+    if (!lastJson.HasMember("arguments")) {
```

Comment:
Couldn't we wrap this:
lastJson.HasMember("arguments") as sth more meaningfull? I presume this means:
weAlreadySentDeltaWithArguments?